File: 02-object-oriented/c/11-structs.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef struct Caballo {
    char* color;
    void (*caminar)(void);
    void (*correr)(void);
    void (*parar)(void);  
} Caballo;


typedef struct Pegaso {
    Caballo base;
    void (*volar)(void);  
} Pegaso;


typedef struct Unicornio {
    Caballo base;
    void (*cornear)(void);  
} Unicornio;


typedef struct Centauro {
    Caballo base;
    void (*hablar)(void);  
} Centauro;


Caballo* crearCaballo() {
    Caballo* c = (Caballo*)malloc(sizeof(Caballo));
    c->color = "blanco";
    c->caminar = NULL;
    c->correr = NULL;
    c->parar = NULL;
    return c;
}


Pegaso* crearPegaso(const char* color) {
    Pegaso* p = (Pegaso*)malloc(sizeof(Pegaso));
    p->base.color = (char*)color;
    p->base.caminar = NULL;
    p->base.correr = NULL;
    p->base.parar = NULL;
    p->volar = NULL;  
    return p;
}


Unicornio* crearUnicornio() {
    Unicornio* u = (Unicornio*)malloc(sizeof(Unicornio));
    u->base.color = "blanco";
    u->base.caminar = NULL;
    u->base.correr = NULL;
    u->base.parar = NULL;
    u->cornear = NULL;  
    return u;
}


Centauro* crearCentauro(const char* color) {
    Centauro* c = (Centauro*)malloc(sizeof(Centauro));
    c->base.color = (char*)color;
    c->base.caminar = NULL;
    c->base.correr = NULL;
    c->base.parar = NULL;
    c->hablar = NULL;  
    return c;
}

void mostrarColor(char* color) {
    printf("%s\n", color);
}

int main() {
    Pegaso* p = crearPegaso("negro");
    Unicornio* u = crearUnicornio();
    Centauro* c = crearCentauro("cafe");

    mostrarColor(p->base.color);
    mostrarColor(u->base.color);
    mostrarColor(c->base.color);

    free(p);
    free(u);
    free(c);

    return 0;
}

File: 03-abstract-data-types/rust/11-structs.rs
#![allow(
    dead_code,
    unused_imports
    )]

trait Alas {
    fn volar(&self);
}

trait Cuerno {
    fn cornear(&self);
}

trait Torso {
    fn hablar(&self);
}

trait Caballo {
    fn caminar(&self);
    fn correr(&self);
    fn parar(&self);
}

struct Pegaso {
    color: String,
}

impl Pegaso {
    fn new(color: &str) -> Self {
        Pegaso {
            color: String::from(color),
        }
    }

    fn show(&self) {
        println!("{}", self.color);
    }
}

impl Caballo for Pegaso {
    fn caminar(&self) {
        println!("El Pegaso está caminando");
    }

    fn correr(&self) {
        println!("El Pegaso está corriendo");
    }

    fn parar(&self) {
        println!("El Pegaso ha parado");
    }
}

impl Alas for Pegaso {
    fn volar(&self) {
        println!("El Pegaso está volando");
    }
}

struct Unicornio {
    color: String,
}

impl Unicornio {
    fn new() -> Self {
        Unicornio {
            color: String::from("blanco"),
        }
    }

    fn show(&self) {
        println!("{}", self.color);
    }
}

impl Caballo for Unicornio {
    fn caminar(&self) {
        println!("El Unicornio está caminando");
    }

    fn correr(&self) {
        println!("El Unicornio está corriendo");
    }

    fn parar(&self) {
        println!("El Unicornio ha parado");
    }
}

impl Cuerno for Unicornio {
    fn cornear(&self) {
        println!("El Unicornio está corneando");
    }
}

struct Centauro {
    color: String,
}

impl Centauro {
    fn new(color: &str) -> Self {
        Centauro {
            color: String::from(color),
        }
    }

    fn show(&self) {
        println!("{}", self.color);
    }
}

impl Caballo for Centauro {
    fn caminar(&self) {
        println!("El Centauro está caminando");
    }

    fn correr(&self) {
        println!("El Centauro está corriendo");
    }

    fn parar(&self) {
        println!("El Centauro ha parado");
    }
}

impl Torso for Centauro {
    fn hablar(&self) {
        println!("El Centauro está hablando");
    }
}

fn main() {
    let p = Pegaso::new("negro");
    let u = Unicornio::new();
    let c = Centauro::new("café");

    p.show();
    u.show();
    c.show();

}
